### PR TITLE
Add new token added event (duplicating the existing event structure) when collectible is manually added

### LINKED ISSUE #13477: Add event metrics on NFT support in extension
### Describe the bug

Notion reference: https://www.notion.so/NFT-support-in-extension-v1-844f8c3e72e044c4b8d6f7da7e24b646#c9bf0bdb95e044378b6b382953e76a20

### Acceptance Criteria

~~1. **Usage of NFTs. Goal: create asset value per user metric.** i.e. NFT: $15 per user, Non-NFT: $80 per user
    a. Assets vs NFTs
        i. Asset types and value
        ii. Number of NFTs and NFT types (floor value if possible)
    b. Number of NFT collections~~
        
nft 1a
nft 1b

        
~~2. **Detection utility. Goal: elevate usage of detection that’ll lead to improvements of detection overall.**
    a. NFT detection allowed vs disabled
        i. Centralized detection
        ii. Non-centralized detection~~
            
nft 2a
            
b. NFTs added via detection
        i. Added via NFT detection
        ii. Added via Token detection
            
nft 2b
nft 2bii

c. Manually added
        
nft 2c

        
3. **Flow analysis. Goal: track flow of user journey to determine drop-offs and exits**
    a. Adding flow
        i. NFT detected (2bii) → NFT detection (centralized (2ai) / non-centralized (2aii)) → Added via NFT detection (2bi)
        ii. Manual add (2c) → Added via manual 
   b. Sending flow
        i. NFT Details
        ii. Address Added 
        iii. NFT Sent
        
nft 3bi
nft 3bii
nft 3biii

### Tasks
- [x] Add new user trait for 'NFT Autodetection Enabled'. (done #14367) 
- [x] Add new user trait for 'OpenSeas API Enabled'. (done #14367) 
- [x] Add new user trait for 'Number of NFTs'. (done #14495)
- [x] Add new user trait for 'Number of Tokens'. (done in #14427)
- [x] Add new user trait for 'Number of NFT collections'. (done #14377)
- [x] Modify existing token added events to record the `asset_type` using `ASSET_TYPES` from transactions const
- [x] Modify existing token added events to record the `token_standard` using `TOKEN_STANDARDS` from 'ui/helpers/constants/common'
- [x] Add new token added event (duplicating the existing event structure) when auto detection occurs for NFTs (in progress #14279)
- [x] Add metric tracking for NFT detection. See `CollectionDetectionController` and acceptance criteria # 3a
   - Set source property to detected on the new token added event (In Progress https://github.com/MetaMask/controllers/pull/814)
- [x] Add metric tracking for NFT send flow. See `CollectionDetectionController` and acceptance criteria # 3b
- [x] Add `asset_type` to transaction events (done in #13858)
- [x] Add `token_standard` to transaction events (done in #13858)

#### Collectibles state help
Collectibles state comes from the CollectiblesController from metamask/controllers. The key in our state tree we'll want to use is `allCollectibles`. So in the handleMetaMaskUpdate function on the MetaMetrics controller introduced in #14192 you will add some new keys to the traits object for the various tasks above specifying add new user trait.

```js
/**
 * @type Collectible
 *
 * Collectible representation
 * @property address - Hex address of a ERC721 contract
 * @property description - The collectible description
 * @property image - URI of custom collectible image associated with this tokenId
 * @property name - Name associated with this tokenId and contract address
 * @property tokenId - The collectible identifier
 * @property numberOfSales - Number of sales
 * @property backgroundColor - The background color to be displayed with the item
 * @property imagePreview - URI of a smaller image associated with this collectible
 * @property imageThumbnail - URI of a thumbnail image associated with this collectible
 * @property imageOriginal - URI of the original image associated with this collectible
 * @property animation - URI of a animation associated with this collectible
 * @property animationOriginal - URI of the original animation associated with this collectible
 * @property externalLink - External link containing additional information
 * @property creator - The collectible owner information object
 * @property isCurrentlyOwned - Boolean indicating whether the address/chainId combination where it's currently stored currently owns this collectible
 */
```

All collectibles has this shape:
```js
 allCollectibles: { [key: string (account)]: { [key: string (chainId)]: Collectible[] } };
```

This is an object whose keys are accounts and the values are objects where the keys are chainIds. So, for example, in order to get an aggregate number of NFTs imported you'll have to go through each chainId in each account in this object and sum the length of the Collectibles array.

In order to get the number of collections you'll need to do the same sort of thing but you'll want the number of unique 'address' properties on the entries. Each collectible has an address property that indicates which collection it is a part of.

### PR DESCRIPTION
**Pull request by segun**
Add new token added event (duplicating the existing event structure) when collectible is manually added

Set source property to custom on the new token added event,

## More information

* Progresses #13477

### CODE REVIEW

File: ui/pages/add-collectible/add-collectible.js
Comment (brad-decker) [inline ui/pages/add-collectible/add-collectible.js:93]: best to leave these empty strings as undefined (just omit the properties), segment will strip them anyways.

**Review by digiwand — COMMENTED** [review]
LGTM after updating the code based on what Brad [commented](https://github.com/MetaMask/metamask-extension/pull/14279#discussion_r838950067)

I see that the reused pattern from the codebase has been designed to support more events 👍 Can we get a reference to the related, event tracking pattern you found in the codebase?

File: ui/pages/add-collectible/add-collectible.js
Comment (segun) [reply ui/pages/add-collectible/add-collectible.js]: Fixed [here](https://github.com/MetaMask/metamask-extension/pull/14279/commits/1fd076232dab0526122622ef0968e9f664518c6d)

File: ui/pages/add-collectible/add-collectible.js
Comment (digiwand) [inline ui/pages/add-collectible/add-collectible.js:95]:
```suggestion
        name: newCollectible.name || null,
```
nit: we could shorten this. Same comment for the `description` prop below

**Review by digiwand — DISMISSED** [review]
LGTM! left just a minor suggestion related to [lines 95 & 96](https://github.com/MetaMask/metamask-extension/pull/14279/files#diff-d1e7bb97394326d6e78a5d8631e4d88531660b2f9e21ce61f5f3906dd556eac5R95)

File: ui/pages/add-collectible/add-collectible.js
Comment (segun) [reply ui/pages/add-collectible/add-collectible.js]: fixed [here](https://github.com/MetaMask/metamask-extension/pull/14279/commits/2dff6f9ebd06242a63bd80eec250ec59d245e5e4)

File: ui/pages/add-collectible/add-collectible.js
Comment (adonesky1) [inline ui/pages/add-collectible/add-collectible.js:73]: might be worth making this a case-insensitive check until we have thoroughly standardized checksumming across controllers.

Comment (segun) [reply ui/pages/add-collectible/add-collectible.js]: fixed [here](https://github.com/MetaMask/metamask-extension/pull/14279/commits/d68dc9564e5ee8b6f7989ccbbb799b3604389567)

File: ui/pages/add-collectible/add-collectible.js
Comment (segun) [reply ui/pages/add-collectible/add-collectible.js]: fixed [here](https://github.com/MetaMask/metamask-extension/pull/14279/commits/d68dc9564e5ee8b6f7989ccbbb799b3604389567)

**Review by digiwand — DISMISSED** [review]
Awesome, LGTM! 

Good to know about [@adonesky1's comment above](https://github.com/MetaMask/metamask-extension/pull/14279#discussion_r842893856):
> might be worth making this a case-insensitive check until we have thoroughly standardized storing addresses as checksummed across controllers.

now I'm wondering if we want to normalize the addresses here and/or through Segment.
cc: @brad-decker

**Review by jpuri — DISMISSED** [review]

File: ui/pages/add-collectible/add-collectible.js
Comment (adonesky1) [inline ui/pages/add-collectible/add-collectible.js:73]: [nit] - you could/should use the `isEqualCaseInsensitive` util method

**Review by brad-decker — CHANGES_REQUESTED** [review]
potentially blocking question: ... Is this tracking token added when the token is detected or when the token is added to the wallet? :

File: ui/pages/add-collectible/add-collectible.js
Comment (adonesky1) [inline ui/pages/add-collectible/add-collectible.js:55]: Should this be collectible/NFT added? Sorry I don't have context on the schema we're using, are we somehow indicating that this is an collectible/NFT elsewhere as a subcategory of token?

File: ui/pages/add-collectible/add-collectible.js
Comment (adonesky1) [inline ui/pages/add-collectible/add-collectible.js:99]: A couple of broader questions adding to @brad-decker's point.
1. This is not a "detected" add token event. This is a manually added token. 
and 2. I'm a bit confused about why this track event is being called in a `useEffect` on the page where the user is adding the collectible. Is the idea that the value of `collectibles` will change when the user submits the form to add the collectible and this effect will happen before the component dismounts from the history.push event called on line 132 (at the end of the submission handler)?

Comment (segun) [reply ui/pages/add-collectible/add-collectible.js:99]: For 2: Yes. I did test this. The useEffect is called before the component dismounts. 
For 1: Do we have a "detected" flow for NFTs?

**Comment by segun** [comment]
> potentially blocking question: ... Is this tracking token added when the token is detected or when the token is added to the wallet? :

It's added when the token is added to the wallet.

File: ui/pages/add-collectible/add-collectible.js
Comment (digiwand) [reply ui/pages/add-collectible/add-collectible.js:99]: oh. interesting. I was also confused by the original request, "when auto detection occurs for NFTs".

The existing patterns for "Token Added" do not live in `useEffect` hooks. If we were following the pattern used for "Token Added", we could call `trackEvent` in the same method or callback that adds the token. In this case, could we call track in `handleAddCollectible`?  I'm wondering where you found this pattern original pattern with `useEffect`

File: ui/pages/add-collectible/add-collectible.js
Comment (brad-decker) [reply ui/pages/add-collectible/add-collectible.js:56]: No we should use the asset_type and token_standard properties that match the transaction events to indicate that this 'Added Token' is a collectible/NFT

File: ui/pages/add-collectible/add-collectible.js
Comment (brad-decker) [inline ui/pages/add-collectible/add-collectible.js:77]: we're missing properties here. asset_type, token_standard... and source should 'custom' (if this is when the user manually types in all the details of the token)

**Review by brad-decker — CHANGES_REQUESTED** [review]

File: ui/pages/add-collectible/add-collectible.js
Comment (segun) [reply ui/pages/add-collectible/add-collectible.js]: fixed [here](https://github.com/MetaMask/metamask-extension/pull/14279/commits/a82b0a8b6ff401b1be075480a26a729d3ba46c24)

File: ui/pages/add-collectible/add-collectible.js
Comment (segun) [reply ui/pages/add-collectible/add-collectible.js]: fixed [here](https://github.com/MetaMask/metamask-extension/pull/14279/commits/06de1bf378dce041c50a35dcf20691fa6e123d49)

File: ui/pages/add-collectible/add-collectible.js
Comment (adonesky1) [inline ui/pages/add-collectible/add-collectible.js:80]: We actually don't need this here since it is not used for non-ERC20 tokens. Its a weird function signature that we will change (to accept an object) but for now you can and should put the second arg as `null` and no longer need to get the selectedAddress from state.

File: ui/pages/add-collectible/add-collectible.js
Comment (adonesky1) [reply ui/pages/add-collectible/add-collectible.js]: You could add this field back now since some NFTs have a symbol and it is returned by the `getTokenStandardAndDetails` method that you are using now.

Comment (adonesky1) [reply ui/pages/add-collectible/add-collectible.js]: cc @segun ☝️

**Comment by adonesky1** [comment]
@segun just want to make sure you see my comment [here](https://github.com/MetaMask/metamask-extension/pull/14279#discussion_r848775987)

**Comment by chimerahedonista** [comment]
> @segun just want to make sure you see my comment [here](https://github.com/MetaMask/metamask-extension/pull/14279#discussion_r848775987)

Oh I missed that comment.

File: ui/pages/add-collectible/add-collectible.js
Comment (chimerahedonista) [reply ui/pages/add-collectible/add-collectible.js]: Somehow missed this comment....fixed [here](https://github.com/MetaMask/metamask-extension/pull/14279/commits/b2b69c05d13ded9b2180dcda2e1afe06c9a004bf)

Comment (segun) [reply ui/pages/add-collectible/add-collectible.js]: Somehow missed that....fixed [here](https://github.com/MetaMask/metamask-extension/pull/14279/commits/b2b69c05d13ded9b2180dcda2e1afe06c9a004bf)

Comment (adonesky1) [reply ui/pages/add-collectible/add-collectible.js]: @segun that commit doesn't see to re-add the `token_symbol` to the event?

**Comment by adonesky1** [comment]
@segun another comment still to address here: https://github.com/MetaMask/metamask-extension/pull/14279#discussion_r850491673

File: ui/pages/add-collectible/add-collectible.js
Comment (adonesky1) [inline ui/pages/add-collectible/add-collectible.js:88]: just a question what does `source: 'custom'` mean?

**Review by adonesky1 — APPROVED** [review]
On non blocking question but LGTM

File: ui/pages/add-collectible/add-collectible.js
Comment (brad-decker) [reply ui/pages/add-collectible/add-collectible.js:88]: custom means its added manually. possible options right now are 'custom' | 'dapp' | 'list' | 'detected'

**Review by digiwand — APPROVED** [review]
👍  can we update the description too please? 

currently, it says: 
> Set source property to manual on the new token added event

and we've updated the code to set the source to "custom"

File: ui/pages/add-collectible/add-collectible.js
Comment (digiwand) [reply ui/pages/add-collectible/add-collectible.js:88]: @segun 

should we set this to `detected`? In [#13477](https://github.com/MetaMask/metamask-extension/issues/13477), one of our tasks is to 
>  Set source property to detected on the new token added event

**Comment by adonesky1** [comment]
@brad-decker have your "requested changes" been addressed? Currently this PR is only blocked by that request.

**Comment by brad-decker** [comment]
me: Welp, this is good to go..why aren't they merging it.
also me: .... 🤦🏻‍♂️

**Review by brad-decker — APPROVED** [review]

**Pull request merged**
